README.md
[RKinetica.build]: <https://github.com/kineticadb/RKinetica/releases>
[RKinetica.docs]: <https://github.com/kineticadb/RKinetica/blob/master/RKinetica-manual.pdf>


# RKinetica Connector Guide

The following guide provides step by step instructions to get started using
*R* with *Kinetica*. The *RKinetica* connector provides access to the *Kinetica*
database through a REST interface. The *RKinetica* package can be built locally
from source or a prebuilt package release can be downloaded from the
[release page][RKinetica.build]. *Kinetica* strongly recommends that release
builds be used instead of building the package locally.

*RKinetica* interfaces with *Kinetica* similarly to other *R*-database
interactions: using the *DBI* and *dplyr* packages. Interacting with
*RKinetica* is easiest using the *RStudio* interface. *RStudio* and *RKinetica*
require that *R* be installed locally.

Source code for the connector can be found at:

* <https://github.com/kineticadb/RKinetica>


## Contents

* [Build & Install](#build--install)
* [Usage](#usage)
* [Configuring High Availablity (HA)](#configuring-high-availablity-ha)
  * [Automatic Discovery](#automatic-discovery)
  * [Manual Configuration](#manual-configuration)
* [Documentation](#documentation)


## Build & Install

*RKinetica* depends on several other *R* packages from the comprehensive *R*
archive network (*CRAN*):

* [DBI](https://cran.r-project.org/web/packages/DBI/index.html)
* [RJSONIO](https://cran.r-project.org/web/packages/RJSONIO/index.html)
* [httr](https://cran.r-project.org/web/packages/httr/index.html)
* [bit64](https://cran.r-project.org/web/packages/bit64/index.html)
* [hms](https://cran.r-project.org/web/packages/hms/index.html)
* [methods](https://cran.r-project.org/web/packages/R.methodsS3/index.html)
* [purrr](https://cran.r-project.org/web/packages/purrr/index.html)
* [dplyr](https://cran.r-project.org/web/packages/dplyr/index.html)

**NOTE:**  If installing the *RKinetica* package in *RStudio*, the dependencies
should be installed automatically prior to *RKinetica* being installed.
Otherwise, all packages can be installed via *RStudio* (or the *R* console) like
so:

```
install.packages(c("DBI", "RJSONIO", "httr", "bit64", "hms", "methods", "purrr", "stats", "dplyr"))
```

If opting to build the *RKinetica* package instead of downloading, the connector
package can be built as follows:

```
git clone https://github.com/kineticadb/RKinetica.git -b release/v7.1 --single-branch
R CMD build RKinetica
```

This sequence produces a `tar.gz` file, which, once installed, is made
available to *R*. The `tar.gz` file is created in the same directory the
`build` command was issued, i.e. not in the `RKinetica` repository. Verify that
`tar.gz` file was created before installing the *RKinetica* package:

```
ls RKinetica*
```

To install the *RKinetica* package via the command line:

```
R CMD install RKinetica_7.1.0.0.tar.gz
```

To install the *RKinetica* package in *RStudio* (or *R* console):

```
> install.packages("/path/to/RKinetica_7.1.0.0.tar.gz", repos = NULL, type = "source")
```

## Usage

Before using *RKinetica*, the package must be loaded:

```
library(RKinetica)
```

Create a *KineticaConnection* object using the `dbConnect()` method,
passing in *Kinetica* URL, user, and password parameters:

```
con <- dbConnect(RKinetica::Kinetica(),
                 url = "http://<kinetica-host>:9191",
                 username = "<user>",
                 password = "<password>")
```

**NOTE:** A parameter cannot be added to an existing *KineticaConnection*
object. Instead, a new *KineticaConnection* object must be created to
properly initialize any functionality enabled by additional parameters.

**IMPORTANT:** If using *RStudio*, you can use the *rstudioapi* package to
instead prompt for username and password:

```
con <- dbConnect(RKinetica::Kinetica(),
                 url = "http://<kinetica-host>:9191",
                 username = rstudioapi::askForPassword("Database username?"),
                 password = rstudioapi::askForPassword("Database password?"))
```

If you expect a result set from your queries to exceed 10,000 rows, set the
`row_limit` parameter value accordingly:

```
con <- dbConnect(RKinetica::Kinetica(),
                 url = "http://<kinetica-host>:9191",
                 username = "<user>",
                 password = "<password>",
                 row_limit = 1000000L)
```

You can then use the *KineticaConnection* object as a regular DBI connection.
Some of the most common commands:

* Print connection info:

  ```
  dbGetInfo(con)
  ```

* List top level database objects and collections with their types:

  ```
  dbListObjects(con)
  ```

* Get a list of top level database tables and collections names:

  ```
  dbListTables(con)
  ```

* Drop a table if it exists:

  ```
  dbRemoveTable(con, "tableA")
  ```

* Check if the table exists:

  ```
  dbExistsTable(con, "tableA")
  ```

* Write a table with 3 columns:

  ```
  dbWriteTable(con, "tableA", data.frame(a = 1L, b = 2L, c = 3.0), row.names = NULL)
  ```

* List tableA fields:

  ```
  dbListFields(con, "tableA")
  ```

* Add records to tableA:

  ```
  dbAppendTable(con, "tableA", data.frame(a = 2L:3L, b = 3L:4L, c = 4.0:5.0), row.names = NULL)
  ```

* Read table into variable:

  ```
  rows <- dbReadTable(con, "tableA")
  print(rows)
  ```

* Disconnect:

  ```
  dbDisconnect(con)
  ```


Large tables, views, or query results can have pagination parameters passed into
the `dbSendQuery` or `dbSendStatement` methods. Assuming that the
connection was established with a row limit of 1,000,000, i.e.
`row_limit = 1000000L`, the following example query extracts all records
from the `acquisition` table sorted by `l_id`. The data is retrieved in
batches of one million records and the `offset` is increased by 1 million each
batch. This loop continues until the `dbSendQuery()` resultset is returned
empty:

```
sql_query <- "SELECT l_id, product_type, term, score FROM acquisition ORDER BY l_id"
offset <- 0L
repeat {
  result <- dbSendQuery(con, sql_query, limit = 1000000L, offset = offset)

  # work with current data.frame provided in result@data

  if (nrow(result@data) > 0) {
    # increase offset to get next page of data
    offset <- offset + 1000000L
  } else {
    # exit pagination loop when data.frame is empty
    break
  }
}
```

Additional code examples are available
[in the "examples" subdirectory](examples/).

### Strings vs Factors

When RKinetica reads a character list into R dataframe it can be converted in to
a factor. This option is controlled by environment property that's read into
`as.data.frame()` parameter `stringsAsFactors`:

```
stringsAsFactors = default.stringsAsFactors()
```

To set environment option `stringsAsFactors` to TRUE or FALSE explicitly, use
the following syntax at the beginning of your R script or once per session:

```
options(stringsAsFactors = FALSE)
```

### Schema naming support

Starting with version 7.1 Kinetica DB stops supporting comma and period characters
in table and column names. Instead user can use `<schema_name>.<table_name>`
notation with only one period character allowed or `<table_name>` notation to
lookup tables in the user's default schema. RKinetica added parameter
`default_schema` to KineticaConnection object that stores user's default schema
name:
```
con@default_schema
[1] "ki_home"
```

All of the table managing functions of KineticaConnection (_dbCreateTable(),
dbAppendTable(), dbReadTable(), dbWriteTable(), dbExistsTable(), and
dbRemoveTable()_) now allow _name_ argument to be passed as a *character* value
such as `some_schema.tableA` (or `tableA`, short for omitting default schema
in `ki_home.tableA` notation) or a *KineticaId* object modelled after *DBI::Id*
class that encapsulates a named vector with `schema` and `table` parameter values
defined separately:
```
id  <- KineticaId(schema = "ki_home", table = "some_table")
show(id)
[1] <KineticaId> schema = ki_home, table = some_table
```

You can also create *KineticaId* object by submitting an unnamed character value
that is going to be parsed into named schema and table pair:
```
id  <- KineticaId("ki_home.some_table")
show(id)
[1] <KineticaId> schema = ki_home, table = some_table
```


## Configuring High Availablity (HA)

### Automatic Discovery

When two or more Kinetica clusters have been configured for an HA ring via
KAgent, RKinetica will automatically discover the additional Kinetica
instance URLs available in the ring. If the connection to the URL of the primary
cluster fails, each additional URL will be tried until a successful connection
is established; every failed connection will result in a warning message. If
all connection attempts fail, an error message will be thrown. Only the URL of
the primary cluster to connect to needs to be specified (via the `url` parameter
in the `dbConnect` method); the URLs for the failover clusters will be retrieved
from the primary cluster upon first connecting to it. The *KineticaConnection*
object has additional parameters to store these failover URLs (via the `ha_ring`
parameter) as well as other connection information:

```
con <- dbConnect(RKinetica::Kinetica(), url = "http://172.123.45.61:9191")
dbGetInfo(con)

$url
[1] "http://172.123.45.61:9191"

$host
[1] "172.123.45.61"

$port
[1] 9191

$ha_enabled
[1] TRUE

$ha_ring
[1] "http://172.123.45.63:9191" "http://172.123.45.61:9191" "http://172.123.45.62:9191"
```

URLs in `ha_ring` list are randomly selected to balance load on secondary URL
instances when the primary URL fails. You can use a `show()` command on a
*KineticaConnection* object at any time to check which URL is being used in the
current connection:

```
show(con)

<KineticaConnection>
HA enabled
Current url: http://172.123.45.62:9191
```

### Manual Configuration

If you want to provide URLs for failover clusters manually, you can do so by
adding the `ha_ring` parameter to the `dbConnect()` method with a
comma-separated list of URIs for the secondary cluster(s):

```
con <- dbConnect(RKinetica::Kinetica(), url = "http://172.123.45.61:9191",
       ha_ring = "http://172.123.45.62:9191,http://172.123.45.63:9191")
show(con)

<KineticaConnection>
Self-provided HA enabled
Current url: http://172.123.45.61:9191
```


## Documentation

Detailed *RKinetica* information including method and function descriptions,
usage examples, and arguments can be found in the
[documentation][RKinetica.docs].


CHANGELOG.md
# RKinetica Connector Changelog

## Version 7.1

### Version 7.1.1.0 -- 2020-09-22

#### Added

-   Support for schema naming of Kinetica DB tables

### Version 7.1.0.0 -- 2020-07-27

#### Changed

-   Version release


## Version 7.0

### Version 7.0.1.3 -- 2020-01-07

#### Fixed

-   An error in JSON deserialization associated with
    handling non-JSON special literals NaN and Infinity.

#### Removed

    Connection option `assume_no_nulls`, since all JSON
    parsing is done through RJSONIO package with built-in
    NULL-handling.

### Version 7.0.1.2 -- 2019-11-14

#### Fixed

-   An error in NULL-handling behavior logic of `FALSE` path
    of `assume_no_nulls` connection option.

### Version 7.0.1.1 -- 2019-11-04

#### Added

-   Connection option `assume_no_nulls` to allow faster
    JSON deserializing of Kinetica result when the expected
    result is not expected to contian NULL values. When the
    flag `assume_no_nulls=TRUE`, JSON parsing uses *purrr*
    library for faster data processing. When flag value is
    `FALSE` (default value), it uses *rjson* to provide
    accurate NULL-handling.

### Version 7.0.1.0 -- 2019-09-21

#### Added

-   High Availability configuration support with options to
    either expect on automatic discovery of backup clusters
    when primary Kinetica DB instance is HA-enabled or to
    configure KineticaConnection manually by adding `ha_ring`
    parameter to initial dbConnect() call.

-   Pagination for queries where results are expected to exceed
    `row_limit` of Kinetica Connection. (Available in
    in dbSendQuery() and dbSendStatement() methods through
    `limit` and `offset` parameters.)

### Version 7.0.0.2 -- 2019-09-11

#### Added

-   Support of SSL connection with self-signed certificates.

### Version 7.0.0.1 -- 2019-06-04

#### Added

-   Configurable row_limit parameter to KineticaConnection object
-   Comprehensive examples folder
-   Improved package documentation

#### Fixed

-   Prepared statement parameter handling

### Version 7.0.0.0 -- 2019-03-15

-   Initial version


R/kineticaDriver.R
#' @include kineticaObject.R

## Constants
.KineticaPkgName <- "RKinetica"
.KineticaVersion <- getNamespaceVersion(.KineticaPkgName)


#' Class KineticaDriver
#'
#' This is the \pkg{RKinetica} implementation of base [DBIDriver-class] class
#' from \pkg{DBI}. Driver object is created with \code{Kinetica()} constructor, which
#' does not need arguments.
#' See "Usage" section for DBI methods of [DBIDriver-class] overriden by \pkg{RKinetica}.
#' KineticaDriver defines operations for validating connection configuration \code{dbCanConnect()},
#' creating connections \code{dbConnect()}, mapping data types \code{dbDataType()},
#' listing active connections \code{dbListConnections()}, destroying connections \code{dbDisconnect()},
#' etc.
#'
#' @keywords internal
#' @docType class
#' @name KineticaDriver-class
#' @aliases KineticaDriver,Kinetica
#' @export
#' @import DBI
#' @import methods
setClass("KineticaDriver",
        contains = c("DBIDriver", "KineticaObject"),
        slots = list(
          dbname = "character",
          driver.version = "character",
          client.version = "character",
          connections = "environment")
)


#' Kinetica()
#'
#' Default constructor for KineticaDriver class
#' @rdname KineticaDriver-class
#' @export
#' @seealso dbConnect,dbCanConnect
#' @examples
#' \dontrun{
#' drv <- RKinetica::Kinetica()
#' d <- Kinetica()
#'}
Kinetica <- function() {
  connections <- new.env()
  new ("KineticaDriver",
       dbname =  "Kinetica",
       driver.version = .KineticaVersion,
       client.version = .KineticaVersion,
       connections = connections)
}

#' @rdname KineticaDriver-class
#' @export
Kinetica()



#' dbUnloadDriver()
#'
#' Unload RKinetica Driver, releasing memory from Driver connections and embedded results
#' @rdname dbUnloadDriver
#' @family KineticaDriver methods
#' @param drv Kinetica driver object
#' @param ... Other arguments omitted in generic signature
#' @export
setMethod("dbUnloadDriver", "KineticaDriver", function(drv, ...) {
  if(length(ls(drv@connections)>0)) {
    rm(list = ls(drv@connections), envir = as.environment(drv@connections), inherits = FALSE)
  }
  rm(drv)
  gc()
  invisible(TRUE)
})


#' dbCanConnect()
#'
#' Method allows to check if provided Kinetica Connection configuration could be used to connect
#' to Kinetica DB instance successfully. It takes all the arguments of dbConnect signature,
#' creates a connection and destroys it upon exit.
#' @rdname dbCanConnect
#' @family KineticaDriver methods
#' @param drv Kinetica driver object
#' @param ... Other arguments omitted in generic signature
#' @export
#' @examples
#' \dontrun{
#' # If you have a Kinetica DB instance running in local docker image,
#' # the minimum set of parameters to connect is url
#' dbCanConnect(Kinetica(), url = "http://localhost:9191")
#' # TRUE
#' # You can also use host and port combination
#' dbCanConnect(Kinetica(), host = "127.0.0.1", port = 9191)
#' # TRUE
#' # Remote instances would also require username and password.
#' dbCanConnect(Kinetica(), host = "127.0.0.1", port = 9191,
#'     username = "username", password = "Pa$$w0rd")
#' # TRUE
#' # You can obscure username and password values by calling RStudio secure dialogue box:
#' dbCanConnect(Kinetica(), host = "127.0.0.1", port = 9191,
#'              username = rstudioapi::askForPassword("Kinetica DB user"),
#'              password = rstudioapi::askForPassword("Kinetica DB password"))
#' # TRUE
#'}
setMethod("dbCanConnect", "KineticaDriver",
  function(drv, ...) {
    tryCatch(
      {
        con <- dbConnect(drv, ...)
        dbDisconnect(con)
        TRUE
      },
      error = function(e) {
        structure (FALSE, reason = conditionMessage(e))
      })
  })

#' dbGetInfo()
#'
#' Retrieves brief KineticaDriver description:
#' dbname, driver version, client version, max number of connections
#' (NA stands for unlimited)
#' @rdname dbGetInfo
#' @family KineticaDriver methods
#' @param dbObj an object derived of [KineticaDriver-class] type
#' @param ... Other arguments omitted in generic signature
#' @export
#' @examples
#' \donttest{
#' drv <- Kinetica()
#' dbGetInfo(drv)
#' # name
#' #  "KineticaDriver"
#' # dbname
#' #  "Kinetica"
#' # driver.version
#' #  "7.1.0.0"
#' # client.version
#' #  "7.1.0.0"
#' # max.connections
#' #  NA
#'}
setMethod("dbGetInfo", "KineticaDriver",
  function(dbObj, ...) {
    list(name = "KineticaDriver",
         dbname =  "Kinetica",
         driver.version = dbObj@driver.version,
         client.version = dbObj@client.version,
         max.connections = NA )
  }
)


#' dbIsValid()
#'
#' Checks if the provided Driver object is valid
#' @family KineticaDriver methods
#' @rdname dbIsValid
#' @param dbObj an object of [KineticaDriver-class] type
#' @param ... Other arguments omitted in generic signature
#' @export
#' @examples
#' \dontrun{
#' drv <- Kinetica()
#' dbIsValid(drv)
#' #  TRUE
#'}
setMethod("dbIsValid", signature("KineticaDriver"),
  function(dbObj, ...) TRUE
)

#' show()
#'
#' RKinetica Driver short printout string
#' @family KineticaDriver methods
#' @rdname show
#' @export
#' @examples
#' \dontrun{
#' drv <- Kinetica()
#' show(drv)
#' #  KineticaDriver 7.1.0.0
#' }
setMethod("show", "KineticaDriver", function(object) {
  cat("<", is(object)[1], " ", .KineticaVersion, ">\n", sep = "")
})


#' dbListConnections()
#'
#' Lists all active connections in KineticaDriver's connection pool
#' invoking show() on KineticaConnection objects
#' @family KineticaDriver methods
#' @rdname dbListConnections
#' @param drv an object of [KineticaDriver-class] type
#' @param ... Other arguments omitted in generic signature
#' @export
setMethod("dbListConnections", "KineticaDriver", function(drv, ...) {
  if (!dbIsValid(drv)) {
    stop("Driver is invalid", call. = FALSE)
  }
  ls(drv@connections)
})


#' dbDataType()
#'
#' Takes an R object or literal and returns a matching SQLtype
#' used when creating Kinetica DB table value to store it
#' @family KineticaDriver methods
#' @param dbObj KineticaDriver object
#' @param obj any data object or literal
#' @param ... Other arguments omitted in generic signature
#' @importFrom bit64 integer64 is.integer64
#' @export
#' @examples
#' \dontrun{
#' dbDataType(Kinetica(), 1L)
#'  "INTEGER"
#' dbDataType(Kinetica(), 1)
#'  "DOUBLE"
#' dbDataType(Kinetica(), "Test string")
#'  "VARCHAR(256)"
#' dbDataType(Kinetica(), Sys.Date())
#'  "DATE"
#' dbDataType(Kinetica(), list())
#'  "BINARY"
#'}
setMethod("dbDataType", "KineticaDriver", function(dbObj, obj, ...) {

  if (is.null(obj)) {
    stop("NULL is not a valid type", call. = FALSE)
  }
  if (is.factor(obj)) return("VARCHAR(256)")
  if (is.data.frame(obj)) return(callNextMethod(dbObj, obj))
  #   # return(vapply(x, dbDataType, FUN.VALUE = character(1), USE.NAMES = TRUE))
  if (is.integer64(obj)) return("LONG")

  if (class(obj)[[1]] == "AsIs") {
    oldClass(obj) <- oldClass(obj)[-1]
    return (dbDataType(dbObj, obj))
  }

  if (class(obj)[[1]] == "Date") return("DATE")
  if (class(obj)[[1]] %in% c("POSIXct", "POSIXt")) return("TIMESTAMP")
  if (class(obj)[[1]] %in% c("hms", "difftime")) return("TIME")

  if (is.integer(obj)) return("INTEGER")
  if (is.logical(obj)) return("TINYINT")
  if (is.raw(obj)) return("BYTES")
  if (is.list(obj)) return("BINARY")

  switch(typeof(obj),
         integer = "INTEGER",
         double = "DOUBLE",
         character = "VARCHAR(256)",
         logical = "TINYINT",
         list = "BINARY",
         raw = "BYTES",
         warning("Unsupported type", call. = FALSE)
  )
})


#' dbConnect()
#'
#' Connect to a Kinetica DB going through the appropriate authentication procedure.
#' You can have multiple connections open and connections can have multiple results attached.
#' This function may be invoked repeatedly assigning its output to different
#' objects.
#' Authentication can be provided by using username/password variables.
#' row_limit param would limit  the number of rows in query results,
#' it's an integer value allowing up to 2,147,483,647 rows. Value literal
#' must be followed by L, as in row_limit = 1000000L, also see example below.
#' Use [dbCanConnect()] to check if a connection can be established.
#' @family KineticaDriver methods
#' @rdname dbConnect
#' @param drv Kinetica driver object
#' @param host character string for Kinetica DB host address
#' @param port integer value for Kinetica DB port
#' @param url character string for Kinetica DB url (protocol + host + port)
#' @param username character string for Kinetica DB username
#' @param password character string for Kinetica DB password
#' @param timeout  integer value for Kinetica DB connection timeout
#' @param ... Other arguments omitted in generic signature
#' @export
#' @examples
#' \dontrun{
#' # If you jave a Kinetica DB instance running in local docker image,
#' # the minimum set of parameters to connect is url
#' dbConnect(Kinetica(), url = "http://localhost:9191")
#' # You can also use host and port combination
#' dbConnect(Kinetica(), host = "127.0.0.1", port = 9191)
#' # Remote instances would also require username and password.
#' dbConnect(Kinetica(), host = "127.0.0.1", port = 9191,
#'   username = "username", password = "Pa$$w0rd", row_limit = 1000000L)
#' con <- dbConnect(Kinetica(), url = "http://localhost:9191")
#' show(con)
#' # KineticaConnection
#' # url: http://localhost:9191
#' dbDisconnect(con)
#'}
setMethod("dbConnect", "KineticaDriver",
  function(drv, host = NULL, port = NULL, url = NULL, username = NULL, password = NULL, timeout = NULL,
           row_limit = NULL, ha_ring = NULL, ...) {
    username <- ifelse(is.null(username), "", username)
    password <- ifelse(is.null(password), "", password)
    timeout <- ifelse(is.null(timeout), 0L, as.integer(timeout))
    row_limit <- ifelse((missing(row_limit) || is.null(row_limit) || !is.numeric(row_limit)),
                          10000L, as.integer(row_limit))

    if (!is.null(host) && !nchar(host)==0) {
      # format url
      url <- .make_url(host = host, port = port)
    } else if (!is.null(url) && !nchar(url)==0) {
      # extract host and port from url
      obj <- .validate_url(url)
      host <- as.character(obj["host"])
      port <- as.integer(obj["port"])
      url <- as.character(obj["url"])
    } else {
      stop("Unsupported Connection configuration, please supply url string or host and port parameters.", call. = FALSE)
    }

    # verify that Kinetica instance is running and connection can be established
    # on error, user would get a detailed error message from HTTP response
    system_properties <- show_system_properties(url, username, password, ha_ring)
    core_version <- system_properties$version.gpudb_core_version
    version <- paste0(strsplit(core_version, "[.]")[[1]][1:4], collapse= ".")
    if (startsWith(version, "6.") || startsWith(version, "5.")) {
      stop(paste0("Unsupported Kinetica DB version ", version, ". RKinetica expects DB version 7.0 or higher."), call. = FALSE)
    }

    ha_ptr <- new.env()
    ha_enabled <- as.logical(system_properties$conf.enable_ha)
    # if flag value is NA, convert it to FALSE
    ha_enabled <- ifelse(is.na(ha_enabled), FALSE, ha_enabled)

    # set boolean flag and label for self-provided HA ring
    if (!is.na(ha_ring) && !is.null(ha_ring) && ha_ring != "") {
      ha_enabled = TRUE
      ha_ptr[["label"]] <- "Self-provided HA"
      ha_ptr[["current"]] <- 1
    } else {
      ha_ring <- system_properties$conf.ha_ring_head_nodes
      # set boolean flag and label for system-provided HA ring
      if (!is.na(ha_ring) && !is.null(ha_ring) && ha_ring != "") {
        ha_ptr[["label"]] <- "HA"
        ha_ptr[["current"]] <- 1
      }
    }

    # permute the HA ring arbitrarily
    ha_ring <- .permute_ha_ring(ha_ring)
    # find position of primary url in the HA ring
    primary_url_idx <- match(url, ha_ring)
    if (is.na(primary_url_idx)) {
      primary_url_idx <- length(ha_ring) + 1
      ha_ring[[primary_url_idx]] <- url
    }
    # set pointer to position of primary url in the HA ring
    ha_ptr[["current"]] <- primary_url_idx

    # get user's default schema
    default_schema <- get_default_schema(url, username, password)

    # create a new connection object
    ptr <- sha1_hash(paste0(url, username, Sys.time()), key = "Kinetica")
    results <- new.env()
    transaction <- new.env()
    conn <- new ("KineticaConnection", drv = drv, host = host, port = port, url = url,
                 username = username, password = password, timeout = timeout, ptr = ptr,
                 db.version = version, results = results, transaction = transaction,
                 row_limit = row_limit, ha_enabled = ha_enabled, ha_ring = ha_ring,
                 ha_ptr = ha_ptr, default_schema = default_schema, ...)
    drv@connections[[conn@ptr]] <- conn

    return (conn)
  }
)

# internal utility functions not exposed to user

.make_url <- function(host = "character", port = "ANY") {
  if (is.null(port) || !is.numeric(port) || port == 0L) {
    port <- 9191L
  }
  url <- paste0("http://", host, ":", port)
}

.validate_url <- function(.url = "character") {
  # remove trailing "/"
  if (endsWith(.url, "/")) {
    len <- nchar(.url)
    .url <- substr(.url, 1, len-1)
  }

  #check protocol
  parts <- strsplit(.url, "://")
  if (length(parts[[1]]) > 1) {
    #extract protocol
    protocol <- parts[[1]][1]
    remainder <- parts[[1]][2]

    if(!startsWith(protocol, "http")) {
      warning(paste("Protocol ", protocol, "is not valid, replacing with default HTTP protocol"))
      protocol <- "http"
    }
  } else {
    warning(paste("No protocol provided, using default HTTP protocol"))
    # no protocol
    protocol <- "http"
    remainder <- .url
  }

  tails <- strsplit(remainder, "/")
  head <- tails[[1]][1]
  # check if port exists by splitting url on ":"
  heads <- strsplit(head, ":")
  if (length(heads[[1]]) == 1) {
    warning(paste("No port provided, using default port 9191"))
    # no port idetified
    host <- head
    port <- "9191"
  } else {
    # host comes before ":"
    host <- heads[[1]][1]
    port <- heads[[1]][2]
  }
  if (length(tails[[1]])>1) {
    # breadcrumbs
    breadcrumbs <- tails[[1]][-1]
    tail_path <- paste0(breadcrumbs, collapse = "/")
    url <- paste0(protocol, "://", host, ":", port, "/", tail_path)
  } else {
    # no breadcrumbs
    url <- paste0(protocol, "://", host, ":", port)
  }
  list(url = url, host = host, port = port)
}

.permute_ha_ring <- function(uris) {
  # Split the string of uris on comma and convert it to vector
  uris <- unlist(strsplit(uris, ","))
  # check that the value is in fact a non-empty vector
  # with more than one member
  if (is.vector(uris) && length(uris)>0) {
    # randomly permute vector values
    sample(uris, length(uris))
  } else {
    # return original object if it's empty
    uris
  }
}
